blog/content/post/2020-07-06-qubesos-fwupd-core.md

```diff
@@ -0,0 +1,191 @@
+
+[![asciicast](https://asciinema.org/a/8ZHBnq5COvqx1LdA1hnScKNfL.svg)](https://asciinema.org/a/8ZHBnq5COvqx1LdA1hnScKNfL)
+
+### qubes-fwupdmgr get-devices
+
+The `get-devices` command shows information about connected devices.
+
+![qubes_fwupdmgr-devices](/img/qfwupd-get-devices.png)
+
+`qubes_fwupdmgr` calls a `fwupdagent get-devices` in the usbVM and dom0.
+The agent provides a JSON output that contains device information. The main
+script parses the output and shows the information in the structured form.
+
+### qubes-fwupdmgr get-updates
+
+The `get-updates` command shows information about possible updates.
+
+![qubes_fwupdmgr-get-updates](/img/qfwupd-get-updates.png)
+
+The main difference between `get-updates` and `get-devices` appears in the older
+version of the `fwupdagent`. 1.3.9 client version has dedicated command
+`fwupdagent get-updates`, which generates information about possible updates in
+the JSON form. 1.2.12 version has no such command. At first, we need to acquire
+devices information from `fwupdagent get-devices`. Updates details are obtained
+in the comparison of the current device version with versions in a release
+list.
+
+### qubes-fwupdmgr update
+
+During the update process, a user gets a list of devices with information about
+possible updates. The user chooses the device number, which will be updated to
+the newest version. Then the tool downloads, validates, and installs firmware
+update `.cab` archive.
+
+![qubes_fwupdmgr-update](/img/qfwupd-update.png)
+
+Downloading a firmware archive and the metadata updates vary in the verification
+process. `fwupd-download-updates.sh` checks the checksum of the archive in the
+updateVM. Then the `fwupd_receive_updates` copies the firmware to dom0 and it
+verifies the checksum once again. The script unpacks the archive and checks the
+signature of the firmware. Also during the system firmware update,
+`qubes_fwupdmgr` verifies DMI details with the information contained in firmware
+metadata. If the device is connected in usbVM, `qubes_fwupdmgr` copies
+the firmware file to sys-usb. The main script initializes
+`fwupd_usbvm_validate`, that verifies the firmware in usbVM before it will be
+installed.
+
+The update process of the device connected in dom0 is shown below.
+
+[![asciicast](https://asciinema.org/a/cv2Iyv10EkF9lqrRtfkVLvGja.svg)](https://asciinema.org/a/cv2Iyv10EkF9lqrRtfkVLvGja)
+
+Here is the update process of the device connected in sys-usb.
+
+[![asciicast](https://asciinema.org/a/v7ZiSG3Xp9fauzwFkMNZgmHgt.svg)](https://asciinema.org/a/v7ZiSG3Xp9fauzwFkMNZgmHgt)
+
+### qubes-fwupdmgr downgrade
+
+The downgrade process is similar to updating. A user chooses the device number
+and the firmware version, that will be installed on the downgraded device. Then
+the tool downloads, validates, and installs firmware downgrade `.cab` archive.
+The firmware installation proceeds with `--allow-older` flag.
+
+The downgrade process is presented in the following video.
```

Comment:
@Asiderr Below we have some image, not video

Reply:
@macpijan Yeah, there should be a video. Fixed in the last commit.